service/controller/resource/instance/delete.go

```diff
@@ -2,10 +2,133 @@ package instance
 
 import (
 	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	capzv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	capzexpv1alpha3 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
 )
 
 // EnsureDeleted is a noop since the deletion of deployments is redirected to
 // the deletion of resource groups because they garbage collect them.
 func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	cluster, err := util.GetClusterFromMetadata(ctx, r.CtrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	azureCluster, err := r.getAzureClusterFromCluster(ctx, cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	credentialSecret, err := r.getCredentialSecret(ctx, *cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.removeSubnetFromAzureCluster(ctx, azureCluster, azureMachinePool.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = r.removeNodePool(ctx, azureMachinePool, credentialSecret)
+	if err != nil {
+		return microerror.Mask(err)
+	}
```

Comment:
Should these be ordered the other way round? Subnet can't be deleted if there are still VMs running on it. It might be that the "delay" here is still not enough and it takes two reconciliation rounds to complete, but from logical point of view I'd may switch. WDYT?

Reply:
Yeah, it makes more sense that way, I'll change it.